main.py

```diff
@@ -30,15 +37,106 @@ class Options:
 
 def find_fns(source: str) -> List[str]:
     fns = re.findall(r'(\w+)\([^()\n]*\)\s*?{', source)
     return [fn for fn in fns if not fn.startswith('PERM')
             and fn not in ['if', 'for', 'switch', 'while']]
 
+@attr.s
+class Candiate(object):
+    ast: ca.FileAST = attr.ib()
+    seed: Optional[int] = attr.ib(default=None)
+    can_randomize: Optional[int] = attr.ib(default=None)
+
+    orig_fn: ca.FuncDef = attr.ib(init=False, default=None)
+    fn_index: int = attr.ib(init=False, default=0)
+    o_file: Optional[str] = attr.ib(init=False, default=None)
+    cur_ast: ca.FileAST = attr.ib(init=False, default=None)
+    score_value: Optional[int] = attr.ib(init=False, default=None)
+    score_hash: Optional[str] = attr.ib(init=False, default=None)
+    _cache_source: Optional[str] = attr.ib(init=False, default=None)
+
+    @staticmethod
+    @functools.lru_cache(maxsize=1)
+    def _cache_start_source(source: str):
+        parser = CParser()
+        ast = parser.parse(source)
+        orig_fn, fn_index = ast_util.find_fn(ast)
+        ast_util.normalize_ast(orig_fn, ast)
+        return orig_fn, fn_index, ast
+
+    @classmethod
+    def from_source(cls, source: str, cparser: CParser, seed: Optional[int] = None) -> 'Candiate':
+        orig_fn, fn_index, ast = cls._cache_start_source(source)
+        p: Candiate = cls(ast=ast, seed=seed)
+        p.orig_fn = orig_fn
+        p.fn_index = fn_index
+        ast_util.normalize_ast(p.orig_fn, ast)
+        p.reset_ast()
+        return p
+
+    def reset_ast(self) -> None:
+        self.ast.ext[self.fn_index] = copy.deepcopy(self.orig_fn)
+        self._cache_source = None
+
+    def randomize_ast(self, randomizer: Randomizer) -> None:
+        if randomizer.random.uniform(0, 1) >= RANDOMIZER_KEEP_PROB:
+            self.reset_ast()
+        randomizer.random.seed(self.seed)
+        randomizer.randomize(self.ast, self.fn_index)
+        self._cache_source = None
+
+    def __enter__(self) -> 'Candiate':
+        return self
+
+    def get_source(self) -> str:
+        if self._cache_source is None:
+            self._cache_source = ast_util.to_c(self.ast)
+        return self._cache_source
+
+    def compile(self, compiler: Compiler) -> bool:
+        self._remove_o_file()
+        source = self.get_source()
+        self.o_file = compiler.compile(source)
+        return self.o_file is not None
+
+    def score(self, scorer: Scorer) -> None:
+        self.score_value, self.score_hash = scorer.score(self.o_file)
+
+    def _remove_o_file(self) -> None:
+        if self.o_file is not None:
+            try:
+                os.remove(self.o_file)
+            except:
+                pass
+
+    def __del__(self) -> None:
+        self._remove_o_file()
+
+class Profiler(object):
+    class StatType(Enum):
+        perm = 1
+        compile = 2
+        score = 3
+
+    def __init__(self):
+        self.time_stats = dict([(x, 0.0) for x in Profiler.StatType])
```

Comment:
`{x: 0.0 for x in Profiler.StatType}`

Reply:
or `defaultdict(float)` even, if we want to support flexible stat types (up to you)